lib/google/gax/api_call.rb

```diff
@@ -45,74 +46,63 @@ class ApiCall
-      # @param exception_transformer [Proc] If an API exception occurs this transformer is given the original exception
-      #   for custom processing instead of raising the error directly.
       #
-      def initialize stub_method, timeout: nil, metadata: nil, retry_policy: nil, params_extractor: nil,
-                     exception_transformer: nil
-        @stub_method           = stub_method
-        @timeout               = timeout
-        @metadata              = metadata
-        @retry_policy          = retry_policy
-        @params_extractor      = params_extractor
-        @exception_transformer = exception_transformer
+      def initialize stub_method
+        @stub_method = stub_method
       end
 
       ##
       # Invoke the API call.
       #
       # @param request [Object] The request object.
-      # @param options [CallOption, Hash] The options for making the API call.
-      # @param block [Proc] The proc to call when the API call is made.
+      # @param options [ApiCall::Options, Hash] The options for making the API call. A Hash can be provided to customize
+      #   the options object, using keys that match the arguments for {ApiCall::Options.new}. This object should only be
+      #   used once.
+      # @param on_response [Proc] A Proc object to format the response object. The Proc should accept response as an
+      #   argument, and return a formatted response object. Optional.
+      # @param on_operation [Proc] A Proc object to provide a callback of the response and operation objects. The Proc
+      #   will be called with both the response and operation objects. Optional.
+      # @param on_stream [Proc] A Proc object to provide a callback for every streamed response received. The Proc will
+      #   be called with the response object. Should only be used on Bidi and Server streaming RPC calls. Optional.
       #
-      def call request, options: nil, &block
-        options = init_call_options options
-
-        apply_params_extractor! request, options
-
+      # @return [Object, Thread] The response object. Or, when `on_stream` is provided, a thread running the callback
+      #   on streamed response is returned.
+      #
+      def call request, options: nil, on_response: nil, on_operation: nil, on_stream: nil
+        # Converts hash and nil to an options object
+        options = ApiCall::Options.new options.to_h if options.respond_to? :to_h
+        block = compose_stream_proc on_stream: on_stream, on_response: on_response
         deadline = calculate_deadline options
 
         begin
-          op = @stub_method.call request, deadline: deadline, metadata: options.metadata, return_op: true
-
-          res = op.execute
-          yield op if block
-          res
+          operation = stub_method.call request, deadline: deadline, metadata: options.metadata, return_op: true, &block
+
+          if on_stream
+            Thread.new { operation.execute }
```

Comment:
It doesn't. Currently streaming requests are not retried and this matches existing behavior.

By returning the thread users can observe if/when the stream has an error or when it closes. I think this is the right approach, as if gives users the most control.

But, if we can think of a way to observe the thread and respond to errors in the GAX layer we can add it. But for things like Pub/Sub we need to know when the stream errors because we have to create the new stream based off the old stream's unsent contents. It's more complicated than anything Gax can do for us.

Reply:
I agree. I don't think we should do anything special here (other than document that they don't interact).